src/box/iproto.cc

```diff
@@ -2950,23 +2960,32 @@ iproto_send_listen_msg(struct evio_service *binary)
 int
 iproto_listen(const char *uri)
 {
-	if (iproto_send_stop_msg() != 0)
-		return -1;
-	evio_service_stop(&tx_binary);
-	if (uri == NULL) {
-		tx_binary.addr_len = 0;
-		return 0;
-	}
+	struct evio_service tmp;
+	memset(&tmp, 0, sizeof(tmp));
 	/*
 	 * Please note, we bind socket in main thread, and then
 	 * listen this socket in all iproto threads! With this
 	 * implementation, we rely on the Linux kernel to distribute
 	 * incoming connections across iproto threads.
 	 */
-	if (evio_service_bind(&tx_binary, uri) != 0)
+	if (uri != NULL && evio_service_bind(&tmp, uri) != 0)
 		return -1;
-	if (iproto_send_listen_msg(&tx_binary) != 0)
+	if (iproto_send_stop_msg() != 0) {
+		evio_service_stop(&tmp);
 		return -1;
+	}
+	if (uri == NULL) {
+		evio_service_stop(&tx_binary);
+		tx_binary.addr_len = 0;
+		return 0;
+	}
+	if (iproto_send_listen_msg(&tmp) != 0) {
```

Comment:
By the way, what if the fiber is cancelled and `cbus_call` here returns an error, but the listen actually succeeds in iproto thread?

I just checked on the master - it crashes https://github.com/tarantool/tarantool/issues/6480. But didn't check your patch.

Reply:
i found and fixed it. But there is some big problem in case when fiber is cancelled or wakeup. cbus_call return error. And we continue to work as if an error occurred, but in reality there is no error and iproto continues to perform the specified function. After my fix there is no SEGFAULT, but what we must do in this case? Maybe @locker has idea?

Reply:
I fixed it, now cbus_call with TIMEOUT_INFINIRY is not interruptable.

Reply:
The fix looks complicated. Maybe, do it in Lua instead? I mean reverting back to the original URI if listen failed with the same private.cfg_set_listen function.

Reply:
Yes it's very good idea